client/src/local_character_controller.rs

```diff
@@ -0,0 +1,290 @@
+use common::{math, proto::Position};
+
+pub struct LocalCharacterController {
+    /// The last extrapolated inter-frame view position, used for rendering and gravity-specific
+    /// orientation computations
+    position: Position,
+
+    /// The up vector relative to position, ignoring orientation
+    up: na::UnitVector3<f32>,
+
+    /// The quaternion adjustment to the character position to represent its actual apparent orientation
+    orientation: na::UnitQuaternion<f32>,
+}
+
+impl LocalCharacterController {
+    pub fn new() -> Self {
+        LocalCharacterController {
+            position: Position::origin(),
+            orientation: na::UnitQuaternion::identity(),
+            up: na::Vector::z_axis(),
+        }
+    }
+
+    /// Get the current position with orientation applied to it
+    pub fn oriented_position(&self) -> Position {
+        Position {
+            node: self.position.node,
+            local: self.position.local * self.orientation.to_homogeneous(),
+        }
+    }
+
+    pub fn orientation(&self) -> na::UnitQuaternion<f32> {
+        self.orientation
+    }
+
+    /// Updates the LocalCharacter based on outside information. Note that the `up` parameter is relative
+    /// only to `position`, not the character's orientation.
+    pub fn update_position(
+        &mut self,
+        position: Position,
+        up: na::UnitVector3<f32>,
+        preserve_up_alignment: bool,
+    ) {
+        if preserve_up_alignment {
+            // Rotate the character orientation to stay consistent with changes in gravity
+            self.orientation = math::rotation_between_axis(&self.up, &up, 1e-5)
+                .unwrap_or(na::UnitQuaternion::identity())
+                * self.orientation;
+        }
+
+        self.position = position;
+        self.up = up;
+    }
+
+    /// Rotates the camera's view by locally adding pitch and yaw.
+    pub fn look_free(&mut self, delta_yaw: f32, delta_pitch: f32, delta_roll: f32) {
+        self.orientation *= na::UnitQuaternion::from_axis_angle(&na::Vector3::y_axis(), delta_yaw)
+            * na::UnitQuaternion::from_axis_angle(&na::Vector3::x_axis(), delta_pitch)
+            * na::UnitQuaternion::from_axis_angle(&na::Vector3::z_axis(), delta_roll);
+    }
+
+    /// Rotates the camera's view with standard first-person walking simulator mouse controls. This function
+    /// is designed to be flexible enough to work with any starting orientation, but it works best when the
+    /// camera is level, not rolled to the left or right.
+    pub fn look_level(&mut self, delta_yaw: f32, delta_pitch: f32) {
+        // Get orientation-relative up
+        let up = self.orientation.conjugate() * self.up;
```

Comment:
Is `conjugate()` the same as `inverse()` here? Can we use that for better readability, if so?

Reply:
Yeah, it's the same. I forgot `inverse()` was an option.